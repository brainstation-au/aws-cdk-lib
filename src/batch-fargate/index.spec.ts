import * as cdk from 'aws-cdk-lib';
import YAML from 'yaml';
import { BatchFargate, BatchFargateProps } from '.';

const createStack = (batchProps: BatchFargateProps) => {
  class MyStack extends cdk.Stack {
    constructor(scope: cdk.App, id: string, props?:cdk.StackProps) {
      super(scope, id, props);

      const vpc = new cdk.CfnParameter(this, 'VPC', {
        type: 'String',
        default: batchProps.vpcId,
      });

      const subnets = new cdk.CfnParameter(this, 'SubnetIds', {
        type: 'List<String>',
        default: batchProps.subnetIds.join(','),
      });

      new BatchFargate(this, 'MyBatch', {
        vpcId: vpc.valueAsString,
        subnetIds: subnets.valueAsList,
        maxvCpus: batchProps.maxvCpus,
        securityGroupIds: batchProps.securityGroupIds,
      });
    }
  }

  const app = new cdk.App();
  return new MyStack(app, 'MyStack');
};

describe('batch-fargate', () => {
  const stack = createStack({
    vpcId: 'vpc-1234',
    subnetIds: ['subnet-123', 'subnet-234'],
    maxvCpus: 4,
    securityGroupIds: ['sg-1234'],
  });
  const template = cdk.assertions.Template.fromStack(stack);

  test('produces correct cloudformation template', () => {
    expect(YAML.stringify(template.toJSON())).toMatchSnapshot();
  });

  test('creates a IAM Role', () => {
    template.resourceCountIs('AWS::IAM::Role', 1);
  });

  test('creates a compute environment', () => {
    template.resourceCountIs('AWS::Batch::ComputeEnvironment', 2);
  });

  test('creates a batch job queue', () => {
    template.resourceCountIs('AWS::Batch::JobQueue', 1);
  });
});
