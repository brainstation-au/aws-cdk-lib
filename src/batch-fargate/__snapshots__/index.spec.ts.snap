// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`batch-fargate produces correct cloudformation template 1`] = `
"Parameters:
  VPC:
    Type: String
    Default: vpc-1234
  SubnetIds:
    Type: List<String>
    Default: subnet-123,subnet-234
  BootstrapVersion:
    Type: AWS::SSM::Parameter::Value<String>
    Default: /cdk-bootstrap/hnb659fds/version
    Description: Version of the CDK Bootstrap resources in this environment,
      automatically retrieved from SSM Parameter Store. [cdk:skip]
Resources:
  MyBatchBatchServiceRole36DD76B9:
    Type: AWS::IAM::Role
    Properties:
      AssumeRolePolicyDocument:
        Statement:
          - Action: sts:AssumeRole
            Effect: Allow
            Principal:
              Service: batch.amazonaws.com
        Version: 2012-10-17
      Description: AWS Batch Service Role
      ManagedPolicyArns:
        - arn:aws:iam::aws:policy/service-role/AWSBatchServiceRole
  MyBatchComputeEnvironment4D8A6583:
    Type: AWS::Batch::ComputeEnvironment
    Properties:
      Type: MANAGED
      ComputeResources:
        MaxvCpus: 4
        SecurityGroupIds:
          - sg-1234
        Subnets:
          Ref: SubnetIds
        Type: FARGATE
      ServiceRole:
        Fn::GetAtt:
          - MyBatchBatchServiceRole36DD76B9
          - Arn
      State: ENABLED
  MyBatchComputeEnvironmentSpot385E54C0:
    Type: AWS::Batch::ComputeEnvironment
    Properties:
      Type: MANAGED
      ComputeResources:
        MaxvCpus: 4
        SecurityGroupIds:
          - sg-1234
        Subnets:
          Ref: SubnetIds
        Type: FARGATE_SPOT
      ServiceRole:
        Fn::GetAtt:
          - MyBatchBatchServiceRole36DD76B9
          - Arn
      State: ENABLED
  MyBatchJobQueueC944D50B:
    Type: AWS::Batch::JobQueue
    Properties:
      ComputeEnvironmentOrder:
        - ComputeEnvironment:
            Ref: MyBatchComputeEnvironmentSpot385E54C0
          Order: 1
        - ComputeEnvironment:
            Ref: MyBatchComputeEnvironment4D8A6583
          Order: 2
      Priority: 1
      State: ENABLED
Rules:
  CheckBootstrapVersion:
    Assertions:
      - Assert:
          Fn::Not:
            - Fn::Contains:
                - - "1"
                  - "2"
                  - "3"
                  - "4"
                  - "5"
                - Ref: BootstrapVersion
        AssertDescription: CDK bootstrap stack version 6 required. Please run 'cdk
          bootstrap' with a recent version of the CDK CLI.
"
`;

exports[`simple-vpc produces correct cloudformation template 1`] = `
"Parameters:
  VPC:
    Type: String
    Default: vpc-1234
  SubnetIds:
    Type: List<String>
    Default: subnet-123,subnet-234
  BootstrapVersion:
    Type: AWS::SSM::Parameter::Value<String>
    Default: /cdk-bootstrap/hnb659fds/version
    Description: Version of the CDK Bootstrap resources in this environment,
      automatically retrieved from SSM Parameter Store. [cdk:skip]
Resources:
  MyBatchBatchServiceRole36DD76B9:
    Type: AWS::IAM::Role
    Properties:
      AssumeRolePolicyDocument:
        Statement:
          - Action: sts:AssumeRole
            Effect: Allow
            Principal:
              Service: batch.amazonaws.com
        Version: 2012-10-17
      Description: AWS Batch Service Role
      ManagedPolicyArns:
        - arn:aws:iam::aws:policy/service-role/AWSBatchServiceRole
  MyBatchComputeEnvironment4D8A6583:
    Type: AWS::Batch::ComputeEnvironment
    Properties:
      Type: MANAGED
      ComputeResources:
        MaxvCpus: 4
        SecurityGroupIds:
          - sg-1234
        Subnets:
          Ref: SubnetIds
        Type: FARGATE
      ServiceRole:
        Fn::GetAtt:
          - MyBatchBatchServiceRole36DD76B9
          - Arn
      State: ENABLED
  MyBatchComputeEnvironmentSpot385E54C0:
    Type: AWS::Batch::ComputeEnvironment
    Properties:
      Type: MANAGED
      ComputeResources:
        MaxvCpus: 4
        SecurityGroupIds:
          - sg-1234
        Subnets:
          Ref: SubnetIds
        Type: FARGATE_SPOT
      ServiceRole:
        Fn::GetAtt:
          - MyBatchBatchServiceRole36DD76B9
          - Arn
      State: ENABLED
  MyBatchJobQueueC944D50B:
    Type: AWS::Batch::JobQueue
    Properties:
      ComputeEnvironmentOrder:
        - ComputeEnvironment:
            Ref: MyBatchComputeEnvironmentSpot385E54C0
          Order: 1
        - ComputeEnvironment:
            Ref: MyBatchComputeEnvironment4D8A6583
          Order: 2
      Priority: 1
      State: ENABLED
Rules:
  CheckBootstrapVersion:
    Assertions:
      - Assert:
          Fn::Not:
            - Fn::Contains:
                - - "1"
                  - "2"
                  - "3"
                  - "4"
                  - "5"
                - Ref: BootstrapVersion
        AssertDescription: CDK bootstrap stack version 6 required. Please run 'cdk
          bootstrap' with a recent version of the CDK CLI.
"
`;
